engine/new_runtime/src/memories/noncontiguous_memory.rs

```diff
@@ -0,0 +1,342 @@
+    RamShard(RamMemory),
+}
+use MemoryShard::*;
+
+/// NonContiguousMemory only works on data which size corresponds to the hash primitive we use. In our case we use it to
+/// store keys hence the size of the data depends on the chosen box provider
+#[derive(Clone)]
+pub struct NonContiguousMemory {
+    shard1: MemoryShard,
+    shard2: MemoryShard,
+    config: NCConfig,
+}
+
+impl LockedMemory for NonContiguousMemory {
+    /// Locks the memory and possibly reallocates
+    fn update(self, payload: Buffer<u8>, size: usize) -> Result<Self, MemoryError> {
+        NonContiguousMemory::alloc(&payload.borrow(), size, self.config.clone())
+    }
+
+    /// Unlocks the memory and returns an unlocked Buffer
+    // To retrieve secret value you xor the hash contained in shard1 with value in shard2
+    fn unlock(&self) -> Result<Buffer<u8>, MemoryError> {
+        let mut data1 = [0u8; NC_DATA_SIZE];
+        sha::SHA256(&self.get_buffer_from_shard1().borrow(), &mut data1);
+        let data = match &self.shard2 {
+            RamShard(ram2) => {
+                let buf = ram2.unlock()?;
+                let x = xor(&data1, &buf.borrow(), NC_DATA_SIZE);
+                x
+            }
+            FileShard(fm) => {
+                let buf = fm.unlock()?;
+                let x = xor(&data1, &buf.borrow(), NC_DATA_SIZE);
+                x
+            }
+        };
+        Ok(Buffer::alloc(&data, NC_DATA_SIZE))
+    }
+}
+
+impl NonContiguousMemory {
+    /// Writes the payload into a LockedMemory then locks it
+    pub fn alloc(payload: &[u8], size: usize, config: NCConfig) -> Result<Self, MemoryError> {
+        if size != NC_DATA_SIZE {
+            return Err(NCSizeNotAllowed);
+        };
+        let random = random_vec(NC_DATA_SIZE);
+        let mut digest = [0u8; NC_DATA_SIZE];
+        sha::SHA256(&random, &mut digest);
+        let digest = xor(&digest, payload, NC_DATA_SIZE);
+
+        let ram1 = RamMemory::alloc(&random, NC_DATA_SIZE)?;
+        let shard1 = RamShard(ram1);
+        let shard2 = match config {
+            RamAndFile => {
+                let fmem = FileMemory::alloc(&digest, NC_DATA_SIZE)?;
+                FileShard(fmem)
+            }
+            FullRam => {
+                let ram2 = RamMemory::alloc(&digest, NC_DATA_SIZE)?;
+                RamShard(ram2)
+            }
+            // Not supported yet TODO
+            _ => {
+                return Err(LockNotAvailable);
+            }
+        };
+
+        Ok(NonContiguousMemory { shard1, shard2, config })
+    }
+
+    fn get_buffer_from_shard1(&self) -> Buffer<u8> {
+        match &self.shard1 {
+            RamShard(ram) => ram.unlock().expect("Failed to retrieve buffer from Ram shard"),
+            _ => unreachable!("{}", IMPOSSIBLE_CASE),
+        }
+    }
+
+    // Refresh the shards to increase security, may be called every _n_ seconds or
+    // punctually
+    #[allow(dead_code)]
+    fn refresh(self) -> Result<Self, MemoryError> {
```

Comment:
```suggestion
    pub fn refresh(self) -> Result<Self, MemoryError> {
```

Shouldn't this method be public to the user?